Add node affinity support

sagor999 opened the pull request:
Hi,

this pull request adds node affinity support. 
Added documentation.
Added e2e test. 
All tests complete successfully locally. 

Please let me know if anything is missing, would like to get this to finish line, as I have seen numerous requests and PR that did not make it. 
Just going to reference them here so that once this is accepted, they all can be closed:
https://github.com/zalando/postgres-operator/pull/975
https://github.com/zalando/postgres-operator/pull/689
https://github.com/zalando/postgres-operator/issues/493
https://github.com/zalando/postgres-operator/issues/420
https://github.com/zalando/postgres-operator/pull/209

sagor999 commented:
Travis-CI failing with this:
```
hack/../pkg is out of date. Please run hack/update-codegen.sh
The command "hack/verify-codegen.sh" exited with 1.
```
I ran locally and it works fine. And I don't see any local changes. 
Do I need to do anything to make CI work with this change?

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -2228,6 +2241,9 @@ func (c *Cluster) generateConnectionPoolerPodTemplate(spec *acidv1.PostgresSpec)
 			ServiceAccountName:            c.OpConfig.PodServiceAccountName,
 			TerminationGracePeriodSeconds: &gracePeriod,
 			Containers:                    []v1.Container{poolerContainer},
+			Affinity: &v1.Affinity{
+				NodeAffinity: &spec.ConnectionPooler.NodeAffinity,
+			},
```
Hm, should we also apply the configured `c.OpConfig.NodeReadinessLabel` here? Like we do in `generatePodTemplate` for db and backup pods.

sagor999 replied:
Makes sense. Fixed!

FxKu commented on e2e/tests/test_e2e.py:
```diff
@@ -680,6 +680,100 @@ def test_taint_based_eviction(self):
+
+        try:
+            # patch current master node with the label
+            print('patching master node: %s'%current_master_node)
```
```suggestion
            print('patching master node: %s', current_master_node)
```
, instead of % ?

sagor999 replied:
Fixed.

FxKu commented:
Can you run `./hack/update-codegen.sh`? Then travis should pass, I think.

sagor999 commented:
@FxKu hm, I just did run that script, but it did not produce any commitable changes. 
```
./hack/update-codegen.sh 
Generating deepcopy funcs
Generating clientset for acid.zalan.do:v1 at github.com/zalando/postgres-operator/pkg/generated/clientset
Generating listers for acid.zalan.do:v1 at github.com/zalando/postgres-operator/pkg/generated/listers
Generating informers for acid.zalan.do:v1 at github.com/zalando/postgres-operator/pkg/generated/informers
```
Which is a bit odd. Do I need to do anything else?

FxKu commented:
Hm, not the first time, I hearing this. Also doesn't work for one of the committers. Anyway, you can take it from here: https://gist.github.com/FxKu/2eaae658c2c2fd2e4d11ccdeb28cb07f

sagor999 commented:
@FxKu CI is now passing. :)

Jan-M commented:
@sagor999 thanks for the PR and that it does include tests.

For now I just want to comment on that in the future we should somehow start or engage a discussion early on on how the Postgres manifest should change. It is not my opinion that mapping through all these API objects is the right way to go.

E.g we could have aligned on a field for node affinity that just takes labels, making the "manifest" slim and not overblown. The type of affinity could have been some default.

`nodeAffinity: "node-type=i3,reserved-for=postgres"` would be something that comes to mind to specify affinity to node labels.

Or in your case simply: `nodeAffinity: "environment=pci"`

I want to hide some complexity of the K8s configuration from the user of our manifests.

sagor999 commented:
@Jan-M That would have been `nodeSelector`, which does support what you are describing, for example: 
```
nodeSelector:
  node-type: i3
  reserved-for: postgres
```
Which internally is mapped into nodeAffinity by kubernetes.
nodeAffinity does give you more flexibility though. 
There were some discussion about deprecating `nodeSelector` from kubernetes API, but just checked and it seems like that idea was shut down. 

So I think we can still support `nodeAffinity` as an advanced feature, and maybe add another PR that would add support for simpler `nodeSelector` instead?

Jan-M commented:
Interesting discussion, from your point are you really looking for the full complexity of node affinity?

I think it is good that the operator creates "node affinity" question should just be what is define on the "user" side. Sometimes it can be helpful to limit the user to the useful options. If in the end we build "affinityComplexity" again we better just support it like you did.

In the end the topic of moving from nodeSelector to nodeAffinity also highlights that this could have been hidden from the user, what was originally nodeSelector in the Postgres manifest could just be translated to affinity.

ReSearchITEng commented:
Hi,
I could not find any nodeSelector parameter either in crd.
Not sure how to make a production deployments done if there is neither nodeSelector nor affinity. We see this is as very important feature as it stops reaching production.

Regarding the nodeSelector vs affinity, I was checking what others are doing:
ElasticSearch: https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-advanced-node-scheduling.html#k8s_node_affinity -> both nodeSelector and affinity (more exactly, it exposes the pod spec)
CouchBase: https://docs.couchbase.com/operator/current/reference-couchbasecluster.html#spec-servers-pod-spec-nodeselector - only nodeSelector, but note they have additional concepts like ServerGroups which brings some extra functionality on top.

As production deployments scenarios become complex relatively quickly, and probably that's one of the reasons the affinity was introduced: more complex logical conditions: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity

If I were to choose which would bring more production support, I would go with affinity.
(as for keeping pg simple: these are optional fields and won't clutter the pg definition when are not required).

FxKu commented:
@sagor999 can to you rebase with the master to resolve the conflict. We did some changes in the e2e tests.

sagor999 commented:
@FxKu done. Do you have an idea as to when you will merge this PR?

FxKu commented on e2e/tests/test_e2e.py:
```diff
@@ -891,6 +891,102 @@ def test_zzz_taint_based_eviction(self):
+                name="acid-minimal-cluster",
+                body=patch_node_affinity_config)
+
+            # bounce replica and check if it has migrated to proper node that has a label
+            k8s.api.core_v1.delete_namespaced_pod("acid-minimal-cluster-1", "default")
+
+            # wait for operator to apply node affinity changes
+            time.sleep(120)
```
doesn't the previous pod trigger a failover? maybe you can use the `assert_failover` moethod then. Instead of using `sleep` we have recently switched to using `eventuallyEqual` and `eventuallyTrue` that retry and potentially finish way sooner than just waiting.

sagor999 replied:
Fixed. Removed hard coded sleep. and removed second sleep, that was not necessary there.

sagor999 commented:
@FxKu updated e2e and removed hard coded sleep from it.

FxKu commented on pkg/cluster/cluster.go:
```diff
@@ -432,6 +432,12 @@ func (c *Cluster) compareStatefulSetWith(statefulSet *appsv1.StatefulSet) *compa
+		needsReplace = true
+		needsRollUpdate = true
+		reasons = append(reasons, "new statefulset's NodeAffinity doesn't match the current one")
+	}
```
This isn't already covered with a [previous](https://github.com/zalando/postgres-operator/blob/9dc5f2427d68150aa91dc5e9a26661d016551825/pkg/cluster/cluster.go#L375) check on affinity?

sagor999 replied:
You are right, it should be covered by it. I will remove those lines.

FxKu commented on pkg/apis/acid.zalan.do/v1/crds.go:
```diff
@@ -569,6 +654,91 @@ var PostgresCRDResourceValidation = apiextv1.CustomResourceValidation{
+												Format: "int32",
+											},
+										},
+										Required: []string{"preference, weight"},
```
can you also move it above `Properties`? I know, it's not alphabetical to `Required`, but improves reability a little.

sagor999 replied:
will do!

FxKu commented on e2e/tests/test_e2e.py:
```diff
@@ -891,6 +891,98 @@ def test_zzz_taint_based_eviction(self):
+                name="acid-minimal-cluster",
+                body=patch_node_affinity_config)
+
+            # bounce replica and check if it has migrated to proper node that has a label
+            k8s.api.core_v1.delete_namespaced_pod("acid-minimal-cluster-1", "default")
+            # wait for pod to be started
+            k8s.wait_for_pod_start('spilo-role=replica')
```
changing the node affinity triggers a rolling update. Thus, we can check here if the failover happens (`asser_failover`) instead of deleting the replica pod.

sagor999 replied:
Will redo that portion!

FxKu commented on e2e/tests/test_e2e.py:
```diff
@@ -891,6 +891,98 @@ def test_zzz_taint_based_eviction(self):
+                namespace="default",
+                plural="postgresqls",
+                name="acid-minimal-cluster",
+                body=patch_node_remove_affinity_config)
```
Removing the affinity is another rolling update. Therefore, we should guarantee that everything is running when we leave the test.. Check other e2e tests like `test_zz_node_readiness_label` or `test_zzz_taint_based_eviction` how it is handled there.

sagor999 replied:
Will update!

FxKu commented on pkg/apis/acid.zalan.do/v1/crds.go:
```diff
@@ -247,6 +247,91 @@ var PostgresCRDResourceValidation = apiextv1.CustomResourceValidation{
+														Format: "int32",
+													},
+												},
+												Required: []string{"preference, weight"},
```
can you also move it above `Properties`? I know, it's not alphabetical order, but improves readability a little.

sagor999 replied:
will do!

FxKu commented on pkg/apis/acid.zalan.do/v1/postgresql_type.go:
```diff
@@ -208,12 +209,13 @@ type PostgresStatus struct {
+	Mode              string          `json:"mode,omitempty"`
+	DockerImage       string          `json:"dockerImage,omitempty"`
+	MaxDBConnections  *int32          `json:"maxDBConnections,omitempty"`
+	NodeAffinity      v1.NodeAffinity `json:"nodeAffinity,omitempty"`
```
Do we need a `NodeAffinity` setting for the pooler, too? @erthalion @Jan-M 
Can we not reuse the affinity defined in the `Postgresql` spec?
If I see it correctly this field isn't even reflected on creation and syncing, e.g. it is not set in `generateConnectionPoolerPodTemplate`.

FxKu replied:
Can you remove the field `NodeAffintity` from pooler API?

sagor999 replied:
Will do!

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -1183,7 +1196,7 @@ func (c *Cluster) generateStatefulSet(spec *acidv1.PostgresSpec) (*appsv1.Statef
 		effectiveRunAsGroup,
 		effectiveFSGroup,
-		nodeAffinity(c.OpConfig.NodeReadinessLabel),
+		nodeAffinity(c.OpConfig.NodeReadinessLabel, &spec.NodeAffinity),
```
as mentioned above, this isn't done for the pooler pod template. Another question would be, if we should apply `c.OpConfig.NodeReadinessLabel`, too?

sagor999 replied:
Is that a question for me or for other members of postgres-operator team?

FxKu commented on pkg/apis/acid.zalan.do/v1/crds.go:
```diff
@@ -247,6 +247,91 @@ var PostgresCRDResourceValidation = apiextv1.CustomResourceValidation{
 							"user": {
 								Type: "string",
 							},
+							"nodeAffinity": {
```
gosh, that's a lot of lines. Is this validation covering all affinity options? Maybe we just cut it short with:
```
"nodeAffinity": {
	Type:                   "object",
	XPreserveUnknownFields: util.True(),
},
```

FxKu replied:
at least, makes it easier to copy it to the CRD yaml [manifest](https://github.com/zalando/postgres-operator/blob/master/manifests/postgresql.crd.yaml) where this part is still missing, I think

sagor999 replied:
@FxKu for some reason, it was just kept trimming extra fields, even with `XPreserveUnknownFields`. That is why I had to expand it almost fully. :(

sagor999 replied:
I just tested it again, and yes, even with `XPreserveUnknownFields` it still trims `unknown` fields and completely breaks node affinity support. Btw, before Nov 17th, it used to work fine.

FxKu replied:
ok, can you remove this part from the validation as it covers the anti_affinity of the pooler

sagor999 replied:
sounds good! will remove

FxKu commented:
@sagor999 did a proper review now. Sorry it took so long and thanks for your patience. Unfortunately, I've found a few more issues.

1. The `NodeAffinity` support for the connection pooler looks only half done. Maybe we better remove it from this PR to get this merged quicker. But, lets wait for opinions from @Jan-M and @erthalion. 
2. e2e tests could still be improved.
3. Do we diff the affinity twice on sync?
4. Extend CRD validation also in the yaml manifests (manifests folder and helm chart). You could also add an example in the [complete](https://github.com/zalando/postgres-operator/blob/master/manifests/complete-postgres-manifest.yaml) postgres manifest. With comments maybe.

Are you fine with these requests?

sagor999 commented:
@FxKu 
1. I would agree on removing connection pooler support for node affinity for now just to make this PR go through. :) 
2. What type of improvements you are looking for?
3. Hm. We shouldn't. What makes you think so? 
4. Will do.

Give me a bit more info about 2 and 3 and I will try to work on those items soon.

FxKu commented:
It's all there in the inline comments ;)

sagor999 commented:
@FxKu oh. haha. for some reason github did not send me notification email about those, so I totally missed them!

sagor999 commented:
@FxKu I think I addressed all issues. But, I think something broke in github actions? It gives this error currently:
```
Error:  "2020-12-06T21:30:13.185Z"  'error from getOptions'
Error: Error: Command failed: git cat-file -p 135a13f18b0a9a96db03f327200b01d8e28cdb23
fatal: Not a valid object name 135a13f18b0a9a96db03f327200b01d8e28cdb23
```
e2e tests pass locally. 

Let me know also what do you want to do about pooler's node affinity.

sagor999 commented:
@FxKu my e2e test completed successfully, but this one failed:
```
======================================================================
FAIL: test_zzzz_cluster_deletion (test_e2e.EndToEndTestCase)
Test deletion with configured protection
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator/timeout_decorator.py", line 81, in new_function
    return function(*args, **kwargs)
  File "/tests/test_e2e.py", line 1106, in test_zzzz_cluster_deletion
    self.eventuallyEqual(lambda: k8s.count_pdbs_with_label(cluster_label), 0, "Pod disruption budget not deleted")
  File "/tests/test_e2e.py", line 42, in eventuallyEqual
    self.assertEqual(y, x, m.format(y))
AssertionError: 1 != 0 : Pod disruption budget not deleted
```
Which doesn't seem related to any of my changes. Is it known that this test might be flaky?
Btw, it runs fine locally.

FxKu commented:
@sagor999 can you remove NodeAffinity from pooler config and validation? See my new inline comments.

sagor999 commented:
@FxKu I think I addressed all issues. Let me know if there is anything else.

sagor999 commented:
@FxKu this test has failed: `test_zz_node_readiness_label`. But it ran fine locally on my machine. Is it known that this test might be flaky? Should I submit empty change to re-trigger tests again?

FxKu commented on pkg/cluster/k8sres_test.go:
```diff
@@ -864,6 +864,31 @@ func testEnvs(cluster *Cluster, podSpec *v1.PodTemplateSpec, role PostgresRole)
 	return nil
 }
 
+func testNodeAffinity(cluster *Cluster, podSpec *v1.PodTemplateSpec) error {
```
can you add a unit test where this function is actually used :smiley:

sagor999 replied:
Facepalm. :rofl: 
Fixed. 
btw, I found there are a bunch of functions like that in that file that are not used in actual tests.

FxKu replied:
Thanks. Yes, we found that, too :smiley:

FxKu commented:
@sagor999 we found that your addition to unit tests is only a function. It's not called from test case (t *testing.T).

sagor999 commented:
@FxKu fix submitted.

Jan-M commented:
:+1:

Jan-M commented:
Thank you @sagor999 for the quick responses to our not so quick responses. This allows us to put it into the 1.6 we try to sum up now.

FxKu commented:
:+1:
